packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1466,6 +1476,68 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
     return SelectionUtils.getResultBasedOnRect(_rect, localPosition);
   }
 
+  TextPosition? _originWordSelectionStart;
+  TextPosition? _originWordSelectionEnd;
+  bool? _selectableContainsOriginWord;
+  SelectionResult _updateSelectionEdgeByWord(Offset globalPosition, {required bool isEnd}) {
+    // Cache the boundaries of the origin word. If the origin word is not in the
+    // current selectable then the _textSelectionEnd will be null.
+    if (_selectableContainsOriginWord == null) {
+      _originWordSelectionStart ??= _textSelectionStart;
+      _originWordSelectionEnd ??= _textSelectionEnd;
+    }
+
+    if (_originWordSelectionStart != null && _originWordSelectionEnd != null) {
+      _selectableContainsOriginWord ??= true;
+    } else {
+      _selectableContainsOriginWord ??= false;
+    }
+
+    _setSelectionPosition(null, isEnd: isEnd);
+    final Matrix4 transform = paragraph.getTransformTo(null);
+    transform.invert();
+    final Offset localPosition = MatrixUtils.transformPoint(transform, globalPosition);
+    if (_rect.isEmpty) {
+      return SelectionUtils.getResultBasedOnRect(_rect, localPosition);
+    }
+    final Offset adjustedOffset = SelectionUtils.adjustDragOffset(
+      _rect,
+      localPosition,
+      direction: paragraph.textDirection,
+    );
+
+    final TextPosition position = paragraph.getPositionForOffset(adjustedOffset);
+    // Check if the original local position is within the rect, if it is not then
+    // we do not need to look up the word boundary for that position. This is to
+    // maintain a selectables selection collapsed at 0 when the local position is
+    // not located inside its rect.
+    final (TextPosition start, TextPosition end)? wordBoundary = !_rect.contains(localPosition) ? null : _getWordBoundaryAtPosition(position);
+    final TextPosition targetPosition = _clampTextPosition(wordBoundary != null ? isEnd ? wordBoundary.$2 : wordBoundary.$1 : position);
+
+    if (_selectableContainsOriginWord!) {
+      // Swap the start and end.
+      if (targetPosition.offset >= _originWordSelectionEnd!.offset) {
```

Comment:
It is hard to read the logic in these block. Can you add check isEnd in the condition? I think it would be noop in the set selection anyway